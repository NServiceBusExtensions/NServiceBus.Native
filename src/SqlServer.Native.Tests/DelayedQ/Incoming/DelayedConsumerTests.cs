using System.Collections.Concurrent;
using System.Linq;
using System.Threading.Tasks;
using NServiceBus.Transport.SqlServerNative;
using ObjectApproval;
using Xunit;
using Xunit.Abstractions;

public class DelayedConsumerTests : TestBase
{
    string table = "DelayedConsumerTests";

    [Fact]
    public async Task Single()
    {
        var database = await LocalDb();

        using (var connection = await database.OpenConnection())
        {
            var manager = new DelayedQueueManager(table, connection);
            await manager.Create();
            await manager.SendData();
            using (var result = manager.Consume().Result)
            {
                ObjectApprover.VerifyWithJson(result.ToVerifyTarget());
            }
        }
    }

    [Fact]
    public async Task Single_nulls()
    {
        var database = await LocalDb();

        using (var connection = await database.OpenConnection())
        {
            var manager = new DelayedQueueManager(table, connection);
            await manager.Create();
            await manager.SendNullData();
            using (var result = manager.Consume().Result)
            {
                ObjectApprover.VerifyWithJson(result.ToVerifyTarget());
            }
        }
    }

    [Fact]
    public async Task Batch()
    {
        var database = await LocalDb();

        using (var connection = await database.OpenConnection())
        {
            var manager = new DelayedQueueManager(table, connection);
            await manager.Create();
            await manager.SendMultipleData();

            var messages = new ConcurrentBag<IncomingDelayedVerifyTarget>();
            var result = manager.Consume(size: 3,
                    action: message => { messages.Add(message.ToVerifyTarget()); })
                .Result;
            Assert.Equal(3, result.Count);
            Assert.Equal(3, result.LastRowVersion);
            ObjectApprover.VerifyWithJson(messages.OrderBy(x => x.Due));
        }
    }

    public DelayedConsumerTests(ITestOutputHelper output) :
        base(output)
    {
    }
}