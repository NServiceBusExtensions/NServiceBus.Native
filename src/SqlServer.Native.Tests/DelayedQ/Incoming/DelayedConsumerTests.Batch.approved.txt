[
  {
    $type: 'IncomingDelayedVerifyTarget',
    Headers: 'headers',
    Due: 'DateTime 1',
    Body: '{}'
  },
  {
    $type: 'IncomingDelayedVerifyTarget',
    Headers: '{}',
    Due: 'DateTime 2'
  },
  {
    $type: 'IncomingDelayedVerifyTarget',
    Headers: 'headers',
    Due: 'DateTime 3',
    Body: '{}'
  }
]